apps/prairielearn/src/tests/lti13.test.ts

```diff
@@ -0,0 +1,248 @@
+    assert.equal(redirectUrl.searchParams.get('scope'), 'openid');
+    assert.equal(redirectUrl.searchParams.get('response_type'), 'id_token');
+    assert.equal(redirectUrl.searchParams.get('response_mode'), 'form_post');
+    assert.equal(
+      redirectUrl.searchParams.get('redirect_uri'),
+      `${siteUrl}/pl/lti13_instance/1/auth/callback`,
+    );
+    assert.ok(redirectUrl.searchParams.get('nonce'));
+    assert.ok(redirectUrl.searchParams.get('state'));
+
+    const redirectUri = redirectUrl.searchParams.get('redirect_uri') as string;
+    const nonce = redirectUrl.searchParams.get('nonce') as string;
+    const state = redirectUrl.searchParams.get('state') as string;
+
+    const keystore = nodeJose.JWK.createKeyStore();
+    const key = await keystore.generate('RSA', 2048, {
+      alg: 'RS256',
+      use: 'sig',
+      kid: 'test',
+    });
+
+    const joseKey = await jose.importJWK(key.toJSON(true) as any);
+    const fakeIdToken = await new jose.SignJWT({
+      nonce,
+      // The below values are based on data observed by Dave during an actual
+      // login with Canvas.
+      'https://purl.imsglobal.org/spec/lti/claim/message_type': 'LtiResourceLinkRequest',
+      'https://purl.imsglobal.org/spec/lti/claim/version': '1.3.0',
+      'https://purl.imsglobal.org/spec/lti/claim/deployment_id':
+        '7fdce954-4c33-47c9-97b4-e435dbbed9bb',
+      // This MUST match the value in the login request.
+      'https://purl.imsglobal.org/spec/lti/claim/target_link_uri': `${siteUrl}//pl/lti13_instance/1/course_navigation`,
+      'https://purl.imsglobal.org/spec/lti/claim/resource_link': {
+        id: 'f6bc7a50-448c-4469-94f7-54d6ea882c2a',
+        title: 'Test Course',
+      },
+      'https://purl.imsglobal.org/spec/lti/claim/roles': [
+        'http://purl.imsglobal.org/vocab/lis/v2/institution/person#Instructor',
+        'http://purl.imsglobal.org/vocab/lis/v2/membership#Instructor',
+        'http://purl.imsglobal.org/vocab/lis/v2/system/person#User',
+      ],
+      'https://purl.imsglobal.org/spec/lti/claim/context': {
+        id: 'f6bc7a50-448c-4469-94f7-54d6ea882c2a',
+        type: ['http://purl.imsglobal.org/vocab/lis/v2/course#CourseOffering'],
+        label: 'TEST 101',
+        title: 'Test Course',
+      },
+      name: 'Test User',
+      email: 'test-user@example.com',
+      'https://purl.imsglobal.org/spec/lti/claim/custom': {
+        uin: '123456789',
+      },
+    })
+      .setProtectedHeader({ alg: 'RS256' })
+      .setIssuer(`http://localhost:${oidcProviderPort}`)
+      .setIssuedAt()
+      .setExpirationTime('1h')
+      .setSubject('a555090c-8355-4b58-b315-247612cc22f0')
+      .setAudience(CLIENT_ID)
+      .sign(joseKey);
+
+    // Run a server to respond to JWKS requests.
+    const app = express();
+    app.get('/jwks', (req, res) => {
+      res.setHeader('Content-Type', 'application/json');
+      res.end(JSON.stringify(keystore.toJSON()));
```

Comment:
I'd probably put an explicit `keystore.toJSON(false)` here, just to be safe.

Reply:
Done in 0d88dc3.